src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -1636,16 +1671,38 @@ public override BoundNode VisitLocalDeclaration(BoundLocalDeclaration node)
                 }
 
                 type = valueType.ToTypeWithAnnotations();
-                _variableTypes[local] = type;
+                updateLocalSymbols(type);
 
                 if (node.DeclaredTypeOpt != null)
                 {
                     SetAnalyzedNullability(node.DeclaredTypeOpt, new VisitResult(valueType, type), true);
                 }
             }
+            else
+            {
+                updateLocalSymbols();
+            }
 
             TrackNullableStateForAssignment(initializer, type, slot, valueType, MakeSlot(initializer));
             return null;
+
+            void updateLocalSymbols(TypeWithAnnotations? updatedType = null)
+            {
+                var updatedContainingSymbol = GetUpdatedContainingSymbol(node.LocalSymbol);
+                if (Symbol.Equals(node.LocalSymbol.ContainingSymbol, updatedContainingSymbol, TypeCompareKind.ConsiderEverything) &&
+                    (updatedType is null || node.LocalSymbol.TypeWithAnnotations.Equals(updatedType.Value, TypeCompareKind.ConsiderEverything)))
+                {
+                    return;
+                }
+
+                // This is possible in error scenarios with dynamic, where overload resolution fails. See DynamicInvocation_OverloadResolutionFailure
+                if (node.InferredType && updatedType is TypeWithAnnotations { SpecialType: SpecialType.System_Void })
+                {
+                    updatedType = node.LocalSymbol.TypeWithAnnotations;
+                }
+
+                SetLocalType(node, node.LocalSymbol, updatedType ?? node.LocalSymbol.TypeWithAnnotations);
```

Comment:
>SetLocalType(node, node.LocalSymbol, updatedType ?? node.LocalSymbol.TypeWithAnnotations); [](start = 16, length = 90)

Is there a guarantee that we are dealing with "var" variable?

Reply:
No longer relevant.

---
In reply to: [327924974](https://github.com/dotnet/roslyn/pull/38769#discussion_r327924974) [](ancestors = 327924974)